src/stdhasher.cr

```diff
@@ -0,0 +1,248 @@
+require "secure_random"
+
+# Hasher usable for `def hash(hasher)` should satisfy protocol:
+#   class MyHasher
+#     def <<(v : Int::Primitive)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(b : Bytes)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(n : Nil)
+#       # mutate
+#       self
+#     end
+#
+#     def <<(v)
+#       v.hash(self)
+#       self
+#     end
+#
+#     # digest returns hashsum for current state without state mutation
+#     # Note: hashsum should implement commutative `+` if MyHasher used
+#     # for `Hasher#hash(hasher)`
+#     def digest
+#     end
+#
+#     def clone_build
+#       with_state_copy do |copy|
+#         yield copy
+#         copy.digest
+#       end
+#     end
+#   end
+
+# StdHasher used as standard hasher in `Object#hash`
+# It have to provide defenense against HashDos, and be reasonably fast.
+# To protect against HashDos, it is seeded with secure random, and have
+# permutation that hard to forge without knowing seed and seeing hash digest.
+#
+# Also it has specialized methods for primitive keys with different seeds.
+struct StdHasher
+  @@seed = StaticArray(UInt32, 7).new{|i| 0_u32}
+  def self.init
+    raise "#{@@seed[0]} #{@@seed[1]}" unless @@seed[0] == 0_u32 && @@seed[1] == 0_u32
+    buf = pointerof(@@seed).as(Pointer(UInt8))
+    SecureRandom.random_bytes(buf.to_slice(sizeof(typeof(@@seed))))
+  end
+
+  private def initialize(@h : Pointer(Impl))
+  end
+
+  def initialize
+    @h = Pointer(Impl).malloc(1, Impl.new(@@seed[0], @@seed[1]))
```

Comment:
@ysbaddaden, it is not much more unsafe, than anything else with presence of `unsafe_as`.

Yes, it should be clearly documented, that `def hash(hasher)` has no right to store hasher anywhere.

Safe variant is to pass hasher by value, and return new version:
```crystal
module Indexable
  def hash(hasher)
    hasher = hasher.raw(size)
    each do |elem|
      hasher = hasher << elem
    end
    hasher
  end
end
```
But it has several downsides:
1. it is obviously more error prone for author of `def hash(hasher)`,
2. LLVM sucks at optimizing "pass struct with 2 uint32 as an argument and return value"
(probably, not already. I checked that an year ago),
3. it is obviously much more expensive for heavy-weight hashes, such as SHA-[123], Blake2, etc.

That is why I decided to combine stack allocation with pass-by-reference.
I believe, it could be revisited when compiler become smart enough.

I've benchmarked prototype: stack allocation vs hasher as a class is 1:10.
I don't want pay this price for hypothetical programmer, who implements `def hash(hasher)` in a wrong way.

If there were no `Hash#hash` that ignores order of keys, then protocol could be:
```crystal
class MyObj
  def hash
    yield @a
    yield @b
  end
end
module Indexable
  def hash
    yield ::Hashing.raw(size)
    each do |elem|
      yield elem
    end
  end
end
```
It will be safe in any mentioned way, but it will be very hard to implement `Hash#hash` with 'order-independent' property.